### DISCUSSION TITLE
can someone help a julia noob with how to send callbacks to a function.

### DISCUSSION THREAD

**Discussion by ghost** [lbilli/Jib.jl#21] [Q&A] [answered]
Hi all
I am SLOWLY working my way through the julia approach ( coming from Python) and finding it exciting and rational.  The Jib.jl code base is excellent for in context worked examples and project layout. I know this is rookie stuff but I really don't get it.

That said I am having real difficulty with the callback functionality.  I understand the asych concept and the posting of a series of requests to IB, it's the RESPONSE mechanism I am having trouble with.  I just don't get it.
My high level view was that I post a  reqMktData request for, say AAPL and INTC and wait for the data to start flowing back through stdout, the buffers are then wrapped ( in my case simple_wrap) and presented to the code logic using an identifier to differentiate between responses.  **So that's my mental model, if it's accurate then I don't know how to send a reqMktData response to a function I want to define to handle it.**

**Comment by lbilli** [comment, marked as answer]
It's useful to remember that in an asynchronous setting, a program is not executed sequentially, _i.e._ in a linear fashion, but functions invocation is driven by external events, in this case response arrivals.

In the case of `reqMktData()`, whatever needs to be done to handle the market data goes into the body of the functions `tickPrice()`, `tickSize()` and the like, of which the user has total control.

An issue that arises in this context is how to pass data around different parts of a program, since it's not possible to use the usual function arguments nor their return values. Different solutions are available, like:
- use external files or database to be accessed by other programs
- use internal data structures that allow communications between functions/tasks/threads such as channels, closures, ...

Here's an example using closures that stores an up-to-date snapshot of ticker prices in `mdata`, which can be accessed and passed around within the program.

As the response handling is done in a separate `Task` started by `start_reader()`, the program is still single threaded and there's no need to worry about race conditions. More on this [here](https://docs.julialang.org/en/v1/manual/asynchronous-programming/#man-asynchronous).
```julia
using Jib

symbols = ["AAPL", "TSLA", "AMD", "INTC"]

# Data structure shared among the main program and the callbacks
mdata = Dict(tkr => Dict{String,Any}() for tkr in symbols)

wrap = Jib.Wrapper(
         # These events are always sent
         error= (id, errorCode, errorString, advancedOrderRejectJson) -> println("Error: $(something(id, "NA")) $errorCode $errorString $advancedOrderRejectJson"),
         nextValidId= (orderId) -> println("Next OrderId: $orderId"),
         managedAccounts= (accountsList) -> println("Managed Accounts: $accountsList"),

         # Market Data specific events
         tickPrice= function(tickerId, field, price, size, attrib)
                      mdata[symbols[tickerId]][field] = price
                    end,

         tickSize= function(tickerId, field, size)
                     mdata[symbols[tickerId]][field] = size
                    end,

         tickGeneric= function(tickerId, tickType, value)
                        mdata[symbols[tickerId]][tickType] = value
                      end,

         tickString= function(tickerId, tickType, value)
                       mdata[symbols[tickerId]][tickType] = value
                     end,

         marketDataType= function(reqId, marketDataType)
                           # Do nothing
                         end,

         tickReqParams= function(tickerId, minTick, bboExchange, snapshotPermissions)
                          # Do nothing
                        end
        )

ib = Jib.connect(4002, 1)
Jib.start_reader(ib, wrap)

for (idx, s) in enumerate(symbols)
  contract = Jib.Contract(symbol=s, secType="STK", exchange="SMART", currency="USD")
  Jib.reqMktData(ib, idx, contract, "", false)
end

# From this point on, mdata is kept up to date under the hood by the reader task
# It can be accessed and passed around
mdata
# or just AAPL bid
mdata["AAPL"]["BID"]
```

**Reply by ghost** [reply]
THANKS!!!! I'm going to "try" to work through this WONDERFUL example using vscode but right now it's not going too well. I think what's messing me up is trying to dump my old ideas about asynch.  My old model was 
start an asych loop # start_reader expecting results in wrap
post a bunch of requests each with it's own io^tag  #Jib.reqMktData with  idx for symbol
wait until an io^tag appears in the buffer and send it to the relevant function for processing and post another request with it's own io^tag
rinse and repeat until told to end.

**Comment by ghost** [comment]
A question about approach.  

Why Dict and not dataframe?

this has REALLY helped me a lot with my julia adventure. Thanks for doing such a great job.

**Reply by lbilli** [reply]
No reason. Just to show a quick example.
You are free to arrenge the data as you like.

**Reply by ghost** [reply]
thank you. It's just that I am coming from python and have stayed away from Dict preferring dataframes. I'm just a little concerned about the frequency of the data and the performance. I'm having a blast running through your approach and learning a lot. thanks so much

**Reply by ghost** [reply]
sorry to keep bugging you with this but it's important I get this right.  In your experience is a Dict a better datastructure than a dataframe to handle the data flow from IB?  My concern is what happens when I add 200 symbols to my list.  I'm trying not to be a pain but this is something I have no experience with. Normally ( in Python) I would just shove it all into an array or maybe send it out using apache. I've no experience with julia.  ALSO what development environment do you use please? I have tried vscode and it's really not my cup of tea.

**Reply by lbilli** [reply]
I use DataFrames as well, depending on the circumstances. They are both performant in Julia.

I use just a text editor and the REPL. Here are some [Workflow Tips](https://docs.julialang.org/en/v1/manual/workflow-tips/) and the highly recommended packages [Revise.jl](https://github.com/timholy/Revise.jl) and [Debugger.jl](https://github.com/JuliaDebug/Debugger.jl).

**Reply by ghost** [reply]
thank you SO much. I am glad to hear I can move off vscode and back to the REPL and a text editor. Thank you for the links I'll certainly take the time to go through them.
